### PR TITLE
Storage Add-ons: Release upsell to all onboarding flows

### PR DESCRIPTION
**Pull request by jeyip**
Related to https://github.com/Automattic/martech/issues/2023

## Proposed Changes

* Show storage add-on upsells to business and WooCommerce plans in all `/start` and `/setup` onboarding flows

## Testing Instructions

* Check out this branch
* Navigate to either a `/start` or `/setup` onboarding flow and follow steps until at the plans page ( Ex. `/start/onboarding-pm` )
* Select either a 50GB storage add-on or 100GB storage add-on for a business or WooCommerce plan
* Note that if a business or WooCommerce plan aren't displayed in the onboarding flow, then we expected to see _no_ storage add-on upsells
* Spot check that the header prices and billing timeframes update appropriately when a storage add-on is selected
* Verify that the storage add-on is added to the checkout cart when a plan upgrade button is clicked
* Repeat for a variety of other flows as catalogued in PCYsg-RDT-p2

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx [hunk 1]
```diff
@@ -91,7 +92,7 @@ const PlansWrapper: React.FC< Props > = ( props ) => {
 		? reduxHideFreePlan && 'plans-blog-onboarding' === plansIntent
 		: reduxHideFreePlan;
 
-	const onSelectPlan = ( cartItems?: MinimalRequestCartProduct[] | null ) => {
+	const onUpgradeClick = ( cartItems?: MinimalRequestCartProduct[] | null ) => {
```
Comment (jeyip) [inline client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx:94]: **Note:**

I switch the name from `onSelectPlan` to `onUpgradeClick` because we will now be processing _both_ a plan and a storage add-on cart item with this function

**Comment by southp** [comment]
@jeyip I've checked the following scenarios and there are a few cases we need to fix :) 

### The classic signup flows:

* ✅/start/domain
* ✅/start/launch-site
* /start/hosting
    * 😨Choose "Try for free" -> in /plans, choose "Upgrade" with a storage add-on picked -> In the checkout, the storage add-on isn't added.
    * 😨Choose "Get Business" in the plans step with a storage add-on picked -> the storage add-on isn't added.
* ✅/start/onboarding-pm

### The Stepper flows:

* ✅/setup/link-in-bio, /setup/newsletter (there is no Business and Commerce, hence no dropdown)
* 📝/setup/newsletter -> vew all plans -> pick a storage add-on -> this is yet to be verified. We'll need a rebase to include this functionality implemented in #83724 :) 
* 😨/setup/design-first -> picking the Business plan together with the storage add-on in "Choosing your plan" doesn't add the storage add-on into the cart.
* 😨 See if picking a storage add-on in the plans step works -> I somehow can't recall which Stepper flow has Business plan and Commerce plan in their plans, so I added them in the hardcoded way [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/hooks/npm-ready/data-store/use-grid-plans.tsx#L210) for the Newsletter flow. Picking the storage add-on in the plans step doesn't work.

**Comment by jeyip** [comment]
> @jeyip I've checked the following scenarios and there are a few cases we need to fix :)

Thanks for checking this @southp. I've been testing with `/setup/link-in-bio-tld` and selecting a custom domain, but I'll check the other flows now 🙂

_Edit:_

I believe I've addressed all of the issues you were seeing James 👍 I've since smoke tested 

- `/start/domain`
- `/start/launch-site`
- `/start/hosting`
- `/start/onboarding-pm`
- `/setup/link-in-bio`
- `/setup/newsletter`
- `/setup/newsletter` -> view all plans in the `/plans` page
- `/setup/link-in-bio-tld`
- `/setup/design-first`

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/sidebar.tsx [hunk 2]
```diff
@@ -101,10 +101,11 @@ const Sidebar = ( { sidebarDomain, siteSlug, submit, goToStep, flow }: SidebarPr
 
 	const { title, launchTitle, subtitle } = getLaunchpadTranslations( flow );
 
-	const { getPlanCartItem, getDomainCartItem } = useSelect(
+	const { planCartItem, domainCartItem, productCartItems } = useSelect(
```
Comment (jeyip) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/sidebar.tsx:104]: **Note:**

I made the switch from returning selectors in this `useSelect` call to returning derived data ( Ex. getPlanCartItem -> getPlanCartItem() ) because we only need the data in this component.

From my understanding, the only time we'd really need access to the selectors is if we have some sort of event callback where we want to invoke them with dynamic arguments

**Comment by jeyip** [comment]
@southp I'm going to smoke test a few more stepper flows on my own tomorrow, but I believe all the failing scenarios are now fixed 😎

File: client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx [hunk 3]
```diff
@@ -103,6 +104,19 @@ const PlansWrapper: React.FC< Props > = ( props ) => {
...
 		}
 
+		const cartItemForStorageAddOn = cartItems?.find(
+			( items ) => items.product_slug === PRODUCT_1GB_SPACE
+		);
+
+		if ( cartItemForStorageAddOn ) {
+			setProductCartItems( [ cartItemForStorageAddOn ] );
+
+			cartItemForStorageAddOn?.extra &&
```
Comment (southp) [inline client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx]: `?` here looks redundant since we've already checked it at line 111 above.

File: client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx [hunk 4]
```diff
@@ -103,6 +104,19 @@ const PlansWrapper: React.FC< Props > = ( props ) => {
...
+				recordTracksEvent( 'calypso_signup_storage_add_on_upgrade_click', {
```
Comment (southp) [inline client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx]: This works, but it'd introduce a chance of diverging from https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-features-main/index.tsx#L343 since they are now the same code duplicated in two places. Could we consolidate them somewhere? e.g. https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/analytics/signup.js would work.

**Comment by southp** [comment]
Thanks for the fix, @jeyip . The code in general LGTM. The only last instance that's still not working for me is to upgrade to Business plan when I'm on the Business trial:

https://github.com/Automattic/wp-calypso/assets/1842898/9f7b0c27-a001-485c-90aa-1fbe401a7c1b

I'm wondering if it's a wider issue for Atomic sites, though; I hope it's not. Once we get this sorted, I think we will be ready to launch :) 

Also, in case it helps, I found the easier way to make myself eligible for the Business trial is to hardcode this value as `true`: https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-features-main/index.tsx#L426

**Comment by jeyip** [comment]
> I'm wondering if it's a wider issue for Atomic sites, though; I hope it's not. Once we get this sorted, I think we will be ready to launch :)

Pinpointed the problem 🙂 Fixing it after dinner

_Edit:_

There was specific upgrade click logic for the business trial page that was overriding default upgrade click logic. Addressed in https://github.com/Automattic/wp-calypso/pull/83771/commits/afcc9f0c34e0d1895fb53a438c13ac58d655dff8 Currently fixing unit tests

_Edit 2:_

Also worth noting that the CTA doesn't show plan pricing when stickied. I'll make an issue for that tomorrow

File: client/my-sites/plans/trials/business-trial-plans/index.tsx [hunk 5]
```diff
@@ -23,11 +28,15 @@ export function BusinessTrialPlans( props: BusinessTrialPlansProps ) {
 
 			const planPath = getPlanPath( upgradePlanSlug ) ?? '';
 
-			const checkoutUrl = isBusinessPlan( upgradePlanSlug )
-				? getTrialCheckoutUrl( { productSlug: planPath, siteSlug } )
+			const cartItemForStorageAddOn = cartItems?.find(
+				( items ) => items.product_slug === PRODUCT_1GB_SPACE
+			);
+
+			const testUrl = isBusinessPlan( upgradePlanSlug )
```
Comment (southp) [inline client/my-sites/plans/trials/business-trial-plans/index.tsx]: Just a quick drive-by comment: is there a specific reason to name this `testUrl`? :D

Comment (jeyip) [reply client/my-sites/plans/trials/business-trial-plans/index.tsx]: 🤦‍♂️  Totally my mistake! It was meant for logging and I completely spaced on switching it back to checkoutUrl. Thanks for catching this -- I'll do a quick sweep of my code for any other oddities 😅

Comment (jeyip) [reply client/my-sites/plans/trials/business-trial-plans/index.tsx]: Fixed in https://github.com/Automattic/wp-calypso/pull/83771/commits/5012539e57161726a6618c3f3f030ef0fc9e283a

File: client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx (hunk 3 above)
Comment (jeyip) [reply client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx]: Fixed in https://github.com/Automattic/wp-calypso/pull/83771/commits/9469b865704985f19993d75409c2efd61ae20a90

After some review and testing, the tracks event invoked by this condition is actually redundant ( and will be handled by logic in PlansFeaturesMain ). Because of this I remove the logic entirely

File: client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx (hunk 4 above)
Comment (jeyip) [reply client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx]: Thanks for the suggestion 

I actually took a closer look and verified this in testing, but the tracks event logic I included here is redundant https://github.com/Automattic/wp-calypso/pull/83771#discussion_r1388460589. Because of this, I remove it entirely

It should be handled by existing logic in https://github.com/Automattic/wp-calypso/blob/b9d11f5e83540a22ecdbf6a237839d684d4c9977/client/my-sites/plans-features-main/index.tsx#L342-L346

**Review by southp — APPROVED** [review]
Thanks for addressing all the cases cleanly, @jeyip. I've rechecked the previously broken cases and now it works as expected. LGTM 🚢
